compiler/rustc_mir_transform/src/match_branches.rs

```diff
@@ -519,3 +529,212 @@ impl<'tcx> SimplifyMatch<'tcx> for SimplifyToExp {
+            let mut lived_stmts: BitSet<usize> =
+                BitSet::new_filled(bbs[target_bb].statements.len());
+            let mut expected_copy_stmt = None;
+            for (statement_index, statement) in bbs[target_bb].statements.iter().enumerate().rev() {
+                let loc = Location { block: target_bb, statement_index };
+                if let StatementKind::Assign(assign) = &statement.kind {
+                    if !assign.1.is_safe_to_remove() {
+                        return None;
+                    }
+                }
+                match &statement.kind {
+                    StatementKind::Assign(box (place, _))
+                    | StatementKind::SetDiscriminant { place: box place, .. }
+                    | StatementKind::Deinit(box place) => {
+                        if place.is_indirect() || borrowed_locals.contains(place.local) {
+                            return None;
+                        }
+                        let live = live.get_or_insert_with(|| {
+                            MaybeTransitiveLiveLocals::new(&borrowed_locals)
+                                .into_engine(tcx, body)
+                                .iterate_to_fixpoint()
+                                .into_results_cursor(body)
+                        });
+                        live.seek_before_primary_effect(loc);
+                        if !live.get().contains(place.local) {
+                            lived_stmts.remove(statement_index);
+                        } else if let StatementKind::Assign(box (
+                            _,
+                            Rvalue::Use(Operand::Copy(src_place)),
+                        )) = statement.kind
+                            && expected_copy_stmt.is_none()
+                            && expected_src_place == src_place
+                            && expected_dest_place == *place
+                        {
+                            // There is only one statement that cannot be ignored that can be used as an expected copy statement.
+                            expected_copy_stmt = Some(statement_index);
+                        } else {
+                            return None;
+                        }
+                    }
+                    StatementKind::StorageLive(_)
+                    | StatementKind::StorageDead(_)
+                    | StatementKind::Nop => (),
+
+                    StatementKind::Retag(_, _)
+                    | StatementKind::Coverage(_)
+                    | StatementKind::Intrinsic(_)
+                    | StatementKind::ConstEvalCounter
+                    | StatementKind::PlaceMention(_)
+                    | StatementKind::FakeRead(_)
+                    | StatementKind::AscribeUserType(_, _) => {
+                        return None;
+                    }
+                }
+            }
+            let expected_copy_stmt = expected_copy_stmt?;
```

Comment:
Should we perform the same analysis on this statement that we do in the single statement branch above?